### ISSUE TITLE
💡 The `EmailAddress` type

### ISSUE THREAD

**Issue by LVMVRQUXL** [kotools/types#586]
## 📝 Description

Like suggested by @MichaelStH in #55, we would like to introduce a new `EmailAddress` type representing an email address on the Internet.

Here's the goal of API that we should provide after completing this topic:

```kotlin
interface EmailAddress {
    override fun equals(other: Any?): Boolean
    override fun hashCode(): Int
    override fun toString(): String
    
    companion object {
        fun create(text: String): EmailAddress
        fun create(text: String, regex: Regex): EmailAddress
        fun createOrNull(text: String): EmailAddress?
        fun createOrNull(text: String, regex: Regex): EmailAddress?
    }
}
```

## ✅ Checklist

- [x] #339 
- [x] #377 
- [x] #394 
- [x] #581 
- [x] #583 
- [x] #582 
- [x] #593 
- [x] #594
- [x] #600 
- [x] #614
- [x] #601
- [x] #369 
- [ ] Close this topic as completed and update tracking issues if present.

The following issues are nice-to-have but not required for completing the minimum viable API of this type:

- #595

**Comment by LVMVRQUXL** [comment]
See https://github.com/kotools/types/discussions/55#discussioncomment-8812680 for improving the API of this type.

**Comment by LVMVRQUXL** [comment]
A solution for providing the pattern used for validating an `EmailAddress` to Kotlin and Java is to declare it as a `String` instead of a `Regex`.

```kotlin
// Before
val EmailAddress.Companion.regex: Regex
// After
val EmailAddress.Companion.regexPattern: String // or EmailAddress.Regex.pattern
```

It can also be declared as a `NotBlankString`, but its stabilization will be blocked by the issue #404.

```kotlin
// Before
val EmailAddress.Companion.regex: Regex
// After
val EmailAddress.Companion.regexPattern: NotBlankString // or EmailAddress.Regex.pattern
```

**Comment by LVMVRQUXL** [comment]
Another idea for improving this type is to make it comparable alphabetically like the `NotBlankString` type.

**Comment by LVMVRQUXL** [comment]
A better approach to validating email addresses is to use a simpler regular expression by default, like `^\S+@\S+\.\S+$` and to provide an overload of the factory functions accepting a custom regular expression.

```kotlin
EmailAddress.create("contact@kotools.org") // uses the default regular expression
EmailAddress.create("contact@kotools.org", Regex("^[a-z]+@[a-z]+\\.[a-z]+$")) // uses the specified regular expression
```

For validating the custom regular expression, the trick is to check that the given pattern matches the default regular expression. If that's the case, then we should consider the customized regular expression valid.

```kotlin
val result: Boolean = "^[a-z]+@[a-z]+\\.[a-z]+$" matches Regex("^\\S+@\\S+\\.\\S+$")
println(result) // true
```

**Comment by LVMVRQUXL** [comment]
> Comment related to #601.

For making the factory functions accepting a custom regular expression available on Kotlin and on Java, we could change their signature for accepting a custom pattern as `Any`. This way, these functions will use the string representation of the specified pattern as regular expression.

```kotlin
fun EmailAddress.Companion.create(text: String, pattern: Any): EmailAddress
fun EmailAddress.Companion.createOrNull(text: String, pattern: Any): EmailAddress?
```

> [!IMPORTANT]
> Due to the unique behavior of the `Regex.toString()` function on Kotlin/JS, it is recommended to use the `Regex.pattern` property instead.

Here are some examples of calling these functions from Kotlin code:

```kotlin
val regex = Regex("^[a-z]+@[a-z]+\\.[a-z]+$")
EmailAddress.create("contact@kotools.org", regex.pattern) // passes
EmailAddress.create(" @kotools.org", regex.pattern) // fails
```

**Comment by LVMVRQUXL** [comment]
> Related issue: #614.

For hiding implementation details, an option is to remove the `EmailAddress.Companion.regexPattern` property from the public API, and indicate the pattern used for validating email addresses directly in the documentation of the factory functions.

Here's an example of documentation for the `EmailAddress.Companion.create(String)` function:

```kotlin
/**
 * Creates an email address from the specified [text], or throws an [IllegalArgumentException]
 * if the [text] doesn't match the following regular expression: `^\S+@\S+\.\S+$`.
 */
fun EmailAddress.Companion.create(text: String): EmailAddress = TODO()
```

**Comment by LVMVRQUXL** [comment]
> Duplicates #616.

For improving the consistency of our API, we could rename the `text` argument of the factory functions as `value`.
Here's an example for the `EmailAddress.Companion.create(String)` function:

```kotlin
fun EmailAddress.Companion.create(value: String): EmailAddress = TODO()
```

**Comment by LVMVRQUXL** [comment]
> Duplicates #617.

It may be interesting to mark the factory functions of this type with the `JvmStatic` function for improving their usage experience on Java.

**Comment by LVMVRQUXL** [comment]
Topics not being a single unit of work, we close this issue as stale.
